README.md
[![Version](https://img.shields.io/cocoapods/v/XCGLoggerNSLoggerConnector.svg?style=flat)](https://cocoapods.org/pods/XCGLoggerNSLoggerConnector)
[![License](https://img.shields.io/cocoapods/l/XCGLoggerNSLoggerConnector.svg?style=flat)](https://cocoapods.org/pods/XCGLoggerNSLoggerConnector)
[![Platform](https://img.shields.io/cocoapods/p/XCGLoggerNSLoggerConnector.svg?style=flat)](https://cocoapods.org/pods/XCGLoggerNSLoggerConnector)

#XCGLoggerNSLoggerConnector
#####By: Markus Winkler
- Twitter: [@CreativeGun](https://twitter.com/creativegun)

###tl;dr
Adds NSLogger support (with images) to XCGLogger.

###Compatibility

XCGLoggerNSLoggerConnector works for iOS projects. It is a Swift library intended for use in Swift projects.

###How to Use

In your AppDelegate, declare a global constant to the default XCGLogger instance.

```Swift
import XCGLogger
import NSLogger
import XCGLoggerNSLoggerConnector

let log: XCGLogger = {
    let log = XCGLogger.defaultInstance()
    log.setup(.Debug, showThreadName: false, showLogLevel: true, showFileNames: true, showLineNumbers: true, writeToFile: nil, fileLogLevel: .Debug)

    // NSLogger support
    // only log to the external window
    LoggerSetOptions(LoggerGetDefaultLogger(), UInt32( kLoggerOption_BufferLogsUntilConnection | kLoggerOption_BrowseBonjour | kLoggerOption_BrowseOnlyLocalDomain ))
    LoggerStart(LoggerGetDefaultLogger())
    log.add(destination: XCGNSLoggerLogDestination(owner: log, identifier: "nslogger.identifier"))

    return log
}()
```

You can also specify to display inline the file & function name (personal preference, makes it slightly easier to read where the call is coming from).
```
    log.add(destination: XCGNSLoggerLogDestination(owner: log, identifier: "nslogger.identifier", addInlineDebugInfo: true))
```

Now all levels accept an UIImage as a parameter and will output the content to the NSLogger window as an image.

``` Swift
log.verbose(image)
log.debug(image)
log.info(image)
log.warning(image)
log.error(image)
log.severe(image)
```

There is a new filter command, "customLabel". You can use that one to add your own label to filter later on in the NSLogger client.
The default value is "image" for image and "string" for string.
Use it to distinguish between different kinds of images or outputs (e.g. database results)
``` Swift
log.customLabel(image, label: "facebook icon")
log.customLabel(String(databaseOutput), label: "Database")
```
**Note**: the logLevel for customLabels is Info (3)


###References

You can find more info on XCGLogger/NSLogger (documentation and setup) here:
XCGLogger: https://github.com/DaveWoodCom/XCGLogger
NSLogger: https://github.com/fpillet/NSLogger  

###Change Log
* **Version 0.4.0**: *(2018/10/02)* - Update for Swift 4.2
* **Version 0.3.2**: *(2017/05/23)* - Updated first parameter of XCGLogger extension functions in order to override correctly the updated XCGLogger version
* **Version 0.3.1**: *(2016/12/08)* - removed 2 compiler warnings
* **Version 0.3.0**: *(2016/12/08)* - Xcode 8.1 support. Upgrade to iOS 9.
* **Version 0.2.0**: *(2016/11/23)* - Update for Swift 3
* **Version 0.1.4**: *(2016/08/23)* - added XCGLogger limit <= 3.3 since any version above will break this connector
* **Version 0.1.3**: *(2016/06/08)* - fixes
* **Version 0.1.2**: *(2016/02/15)* - added a switch to display additional debug info inline
* **Version 0.1.1**: *(2015/11/20)* - Initial Release



CHANGELOG.md
#Change Log
* **Version 0.4.0**: *(2018/10/02)* - Update for Swift 4.2
* **Version 0.3.2**: *(2017/05/23)* - Updated first parameter of XCGLogger extension functions in order to override correctly the updated XCGLogger version
* **Version 0.3.1**: *(2016/12/08)* - removed 2 compiler warnings
* **Version 0.3.0**: *(2016/12/08)* - Xcode 8.1 support. Upgrade to iOS 9.
* **Version 0.2.0**: *(2016/11/23)* - Update for Swift 3
* **Version 0.1.4**: *(2016/08/23)* - added XCGLogger limit <= 3.3 since any version above will break this connector
* **Version 0.1.3**: *(2016/06/08)* - fixes
* **Version 0.1.2**: *(2016/02/15)* - added a switch to display additional debug info inline
* **Version 0.1.1**: *(2015/11/20)* - Initial Release


XCGLoggerNSLoggerConnector/XCGLoggerNSLoggerConnector.swift
//
//  XCGLoggerSupport.swift
//  adds support for NSLogger in XCGLogger
//
//  add the line below to the init code in appDelegate
//  log.addLogDestination(XCGNSLoggerLogDestination(owner: log, identifier: "nslogger.identifier", addInlineDebugInfo: false))
//
//  New custom image level, accepts UIImage
//  Label is default set to "image", can be used to indicate e.g. source ("facebook", "imgur")
//  -- log.image(image)
//  -- log.image(image, label: "facebook")
//
//  All report levels support now UIImage
//  -- log.info(image)
//
//  New custom text report level to group functional reports
//  log.customLabel("Userdata: \(db.user)", label: "Database")
//
//  Created by Markus on 11/17/15.
//  Copyright © 2015 Markus Winkler. All rights reserved.
//

import XCGLogger
import NSLogger

public class XCGNSLoggerLogDestination: BaseDestination {
    
    // use this variable to show file/functioname in the same line with the message
    // (just a personal preference, let me see the flow of the application better)
    public var addInlineDebugInfo: Bool = false
    
    // Report levels are different in NSLogger (0 = most important, 4 = least important)
    // XCGLogger level needs to be converted to use the bonjour app filtering in a meaningful way
    private func convertLogLevel(_ level:XCGLogger.Level) -> Int32 {
        switch(level) {
        case .severe:
            return 0
        case .error:
            return 1
        case .warning:
            return 2
        case .info:
            return 3
        case .debug:
            return 4
        case .verbose:
            return 5
        case .none:
            return 3
        }
        let x:FileDestination!
    }
    
	public init(owner: XCGLogger, identifier: String, addInlineDebugInfo: Bool = false, outputLevel: XCGLogger.Level = .debug) {
        super.init(owner: owner, identifier: identifier)
        self.addInlineDebugInfo = addInlineDebugInfo
		self.outputLevel = outputLevel
    }
    
    open override func output(logDetails: LogDetails, message: String) {
        
        switch(logDetails.level) {
        case .none:
            return
        default:
            break
        }
        
        if addInlineDebugInfo {
            LogMessageF_va(logDetails.fileName, Int32(logDetails.lineNumber), logDetails.functionName, logDetails.level.description, Int32(convertLogLevel(logDetails.level)), "[\(NSString(string: logDetails.fileName).lastPathComponent):\(logDetails.lineNumber)] -> \(logDetails.functionName) : \(logDetails.message)", getVaList([]))
        } else {
            LogMessageF_va(logDetails.fileName, Int32(logDetails.lineNumber), logDetails.functionName, logDetails.level.description, Int32(convertLogLevel(logDetails.level)), logDetails.message, getVaList([]))
        }
        
    }
}

public extension XCGLogger {

    // declared here again for performance reasons
    private func convertLogLevel(_ level:XCGLogger.Level) -> Int32 {
        switch(level) {
        case .severe:
            return 0
        case .error:
            return 1
        case .warning:
            return 2
        case .info:
            return 3
        case .debug:
            return 4
        case .verbose:
            return 5
        case .none:
            return 3
        }
    }

    public func sendImageToNSLogger(_ image: UIImage?, level: Level, label: String, functionName: String = #function, fileName: String = #file, lineNumber: Int = #line) {
        // check if image is valid, otherwise display error
        if let image: UIImage = image {
            #if swift(>=4.2)
                let png = image.pngData()
            #else
                let png = UIImagePNGRepresentation(image)
            #endif
            
            LogImageDataF(fileName, Int32(lineNumber), functionName, label, convertLogLevel(level), Int32(image.size.width), Int32(image.size.height), png)
            self.logln(level, functionName: functionName, fileName: fileName, lineNumber: lineNumber, closure: {return "Image: \(image)"})
        }
        else {
            self.logln(level, functionName: functionName, fileName: fileName, lineNumber: lineNumber, closure: {return "Invalid Image: \(String(describing: image))"})
        }
    }

    public func customLabel(_ closure: @autoclosure () -> UIImage?, label: String = "image", functionName: String = #function, fileName: String = #file, lineNumber: Int = #line) {
        // check if image is valid, otherwise display error
        var arr = fileName.components(separatedBy: "/")
        var fileName2 = fileName
        if let last = arr.popLast() {
            fileName2 = last
        }
        let level = XCGLogger.Level.none
        if let image: UIImage = closure() {
            #if swift(>=4.2)
                let png = image.pngData()
            #else
                let png = UIImagePNGRepresentation(image)
            #endif

            LogImageDataF(fileName, Int32(lineNumber), functionName, label, convertLogLevel(level), Int32(image.size.width), Int32(image.size.height), png)
            LogMessageF_va(fileName, Int32(lineNumber), functionName, label, Int32(convertLogLevel(level)), "\(image)", getVaList([]))
            self.logln(level, functionName: functionName, fileName: fileName2, lineNumber: lineNumber, closure: {return "Image: \(image)"})
        }
        else {
            self.logln(level, functionName: functionName, fileName: fileName2, lineNumber: lineNumber, closure: {return "Invalid Image: \(String(describing: closure()))"})
        }
    }

    public func customLabel(_ closure: @autoclosure () -> String?, label: String = "string", functionName: String = #function, fileName: String = #file, lineNumber: Int = #line) {
        let level = XCGLogger.Level.none
        var arr = fileName.components(separatedBy: "/")
        var fileName2 = fileName
        if let last = arr.popLast() {
            fileName2 = last
        }

        if let message = closure() {
            LogMessageF_va(fileName, Int32(lineNumber), functionName, label, Int32(convertLogLevel(level)), "\(message)", getVaList([]))
            self.logln(level, functionName: functionName, fileName: fileName2, lineNumber: lineNumber, closure: {return "[\(label)] \(message)"})
        }
        else
        {
            LogMessageF_va(fileName, Int32(lineNumber), functionName, label, Int32(convertLogLevel(level)), "nil", getVaList([]))
            self.logln(level, functionName: functionName, fileName: fileName2, lineNumber: lineNumber, closure: {return "[\(label)] nil"})
        }
    }

    public func verbose(_ closure: @autoclosure () -> UIImage?, functionName: String = #function, fileName: String = #file, lineNumber: Int = #line) {
        let level = XCGLogger.Level.verbose
        sendImageToNSLogger(closure(), level: level, label: level.description, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
    }

    public func debug(_ closure: @autoclosure () -> UIImage?, functionName: String = #function, fileName: String = #file, lineNumber: Int = #line) {
        let level = XCGLogger.Level.debug
        sendImageToNSLogger(closure(), level: level, label: level.description, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
    }

    public func info(_ closure: @autoclosure () -> UIImage?, functionName: String = #function, fileName: String = #file, lineNumber: Int = #line) {
        let level = XCGLogger.Level.info
        sendImageToNSLogger(closure(), level: level, label: level.description, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
    }

    public func warning(_ closure: @autoclosure () -> UIImage?, functionName: String = #function, fileName: String = #file, lineNumber: Int = #line) {
        let level = XCGLogger.Level.warning
        sendImageToNSLogger(closure(), level: level, label: level.description, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
    }

    public func error(_ closure: @autoclosure () -> UIImage?, functionName: String = #function, fileName: String = #file, lineNumber: Int = #line) {
        let level = XCGLogger.Level.error
        sendImageToNSLogger(closure(), level: level, label: level.description, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
    }

    public func severe(_ closure: @autoclosure () -> UIImage?, functionName: String = #function, fileName: String = #file, lineNumber: Int = #line) {
        let level = XCGLogger.Level.severe
        sendImageToNSLogger(closure(), level: level, label: level.description, functionName: functionName, fileName: fileName, lineNumber: lineNumber)
    }

}
